### DISCUSSION TITLE
Decorations being treated as items instead of structures (+other sprite depth issues)

### DISCUSSION THREAD

**Discussion by NotWendy** [FakeFishGames/Barotrauma#13640] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

The files attached to this issue feature all decorative entities which suffer from some kind of sprite depth related issues. There are two big categories to these.

1. Structures which have the same default sprite depth as background walls when placed in the submarine editor. Calendar, one of the background doors, a lot of graffiti sprites. This is just an inconvenience, which means that if you place one of those items on a background wall and don't bother to check sprite depth, it being visible or hidden behind the wall in game is up to a z-fighting coinflip.
2. Decorative structures that were added into the game as **items**. These don't disappear alongside all other decorations when the Structures layer is disabled in the submarine editor. Their sprite depth also cannot go above 900, compared to the max sprite depth of 999 for any normal structure. Affects most of the 1.0 decorations, even though some of them were properly added into the game as structures.

Example of an outpost module with a bunch of decorations still being visible when only items are supposed to be shown. Some of these understandably are treated as items because they have light sources attached to them. Others (curtains, rust, mannequin, wires, non-glowing banners) should ideally be treated as structures.
![ClownModule](https://github.com/FakeFishGames/Barotrauma/assets/47554589/f9432888-2cb4-48eb-922f-95194f9fb3bd)

Full list of items affected by the two issues I mentioned:
![SpriteDepthIssues](https://github.com/FakeFishGames/Barotrauma/assets/47554589/8220d232-0e75-4e30-b29c-e0f44b5fb01e)

Same thing as a .sub file, where you can check the name of every item:
[Sprite Depth Issues.zip](https://github.com/FakeFishGames/Barotrauma/files/14376140/Sprite.Depth.Issues.zip)

### Reproduction steps

For structures with badly chosen default sprite depth - simply place them in the editor to see that the depth is 980, same as that of any background wall.

For decorations that are treated as items - place them in the editor and observe how the default sprite depth they start out with does not work properly if it is above 900.
If you try to manually adjust the sprite depth of any of those items to any number above 900 after placing it, it will get automatically reset to 900, a limit enforced by the decoration being an item.

You can also toggle the different layers in the .sub file I provided to see that the majority of the decorations placed there only disappear when items are off.

### Bug prevalence

Happens every time I play

### Single player or multiplayer?

Happens outside single player or multiplayer game modes (e.g. game launches on startup, something broken in the main menu)

### Version

v1.2.8.0 (Winter Update hotfix 2)

### Which operating system did you encounter this bug on?

Windows

**Comment by NotWendy** [comment]
One more thing, since it's related to the topic. Here's an image which displays the inconsistency in the default sprite depth of hatches. It also demonstrates how the default depth of ladders being the same or less than that of hatches leads to the ladders clipping into hatches, instead of smoothly sliding into them, as the artist who did those sprites probably intended.

I think having the default depth of hatches be 510 and the depth of ladders be 511 would work well.

![Hatches](https://github.com/FakeFishGames/Barotrauma/assets/47554589/793e1fd9-5138-4786-9551-c656999b5d61)

Additional point:
There's a secondary issue at play here. The "frame" of the hatch is rendered on top of the character while they're climbing a ladder in the examples below. (Ladder depth is 511 in both cases.) That looks okay. Ladder is an animation controller and puts the character at its own depth, behind the frame of the hatch.

But if the character sticks through the hatch without using the ladder and the depth of the hatch is greater than 500 (sprite depth of a human), then the frame of the hatch is rendered behind the character. (Top right example.)
That's an inconsistency we can solve by making the hatch's frame always render on top of humans. (Bottom example.)
But if we do that, we face a problem of hatch's buttons being at the same depth as its frame, also rendering on top of the human character. (Bottom right example.)

This can be solved by moving the buttons to decorative sprites and making them a part of the assembly, placing them at greater depth than the frame of the hatch. That's how it was done for doors, but hatches for some reason were created differently.
![image](https://github.com/FakeFishGames/Barotrauma/assets/47554589/30352023-8406-4815-b6e7-2bb63274ea64)

Also, if the "frame" of the hatch is a supporting structure that's implied to be in front of everything, shouldn't the hatch look like this:
![image2](https://github.com/FakeFishGames/Barotrauma/assets/47554589/dc1753e0-71d7-4dda-b153-7c9ce36a83e4)

**Comment by NotWendy** [comment]
Onto an issue that affects every structure and item in the game.
Giving anything a sprite depth of less than 501 makes it render on top of solid walls. For example, if a wall (Shell A, inside wall, shuttle shell) has a sprite depth of 100 and a periscope attached to it has a depth of 400, the periscope will render on top of the wall. Example, visualized:
![Periscopes](https://github.com/FakeFishGames/Barotrauma/assets/47554589/a8a6ea4b-7e29-4b17-9c68-17cce4cdd668)

Periscopes, turrets, hardpoints, loaders, depth charge tubes, wall caps, and most importantly, STAIRS are some of the most impactful examples of this unexplainable behavior.
You can't make stairs render behind the floor while still allowing them to render in front of humans, which is very restrictive.
![Stairs](https://github.com/FakeFishGames/Barotrauma/assets/47554589/bc1b4a01-2b1b-45d9-8cc2-5b3627a93ef1)

Fixing this behavior for everything in the game would break pretty much every submarine, outpost, and any other structure, because people have been relying on this weird system for years. Fixing it for periscopes, turrets, and stairs could be okay. Wall caps too, maybe? People leave them at their standard depth of 30 anyways.

**Comment by ubertpendragon** [comment]
I had the same issue when editing a mod and "converted" the items into structures, but this breaks any .sub using those items

![image](https://github.com/FakeFishGames/Barotrauma/assets/73229309/48549549-808a-45bf-829c-a58dea531cd6)


If items were to be converted into structure the game would somehow have to account for that

**Comment by bananaman6213** [comment]
Regarding the 0.900 depth limit on items:
![depth](https://github.com/user-attachments/assets/6c8705b7-fc46-41a5-a9fd-54254de69cfb)
This code in MapEntity.cs may be responsible. Is this an intentional restriction? I can't imagine why if so.

**Reply by NotWendy** [reply]
Likely intentional, to avoid accidentally hiding actual items (gameplay objects) behind background decorations. It's an okay idea, as long as nobody implements a ton of decorative structures as items in a rushed update.

**Reply by Regalis11** [reply]
It is intentional, but the intention is not to prevent hiding items behind decorations. It's somewhat of a technical thing: the submarine is rendered in a couple of passes, with the first pass being a "background layer" that's visible through the LOS effect. Ensuring there are no items in that layer makes sure you can't e.g. magically see items moving through walls.

**Reply by ubertpendragon** [reply]
It's a problem for any item put outside submarines because if two submarines or a submarine and a beacon come close you'll see items moving through the subs. 

Especially visible with these when fighting pirates
![image](https://github.com/user-attachments/assets/46b4ab5c-1015-4d4d-8100-6f4fa08fb487)

Or beacon "antennas" (which currently serve no purpose)

And Rust that is clearly a decoration but was added as an item instead of a structure (because that's the only way it could be rotated before?)
